src/storage/transaction/TransactionManager.cpp

```diff
@@ -0,0 +1,521 @@
+constexpr int kCommitOutEdgeFail = 29;
+constexpr int kRemoveLockFail = 37;
+#endif
+
+namespace nebula {
+namespace storage {
+
+/*
+ * edgeKey : thrift data structure
+ * rawKey  : NebulaKeyUtils::edgeKey
+ * lockKey : rawKey + lock suffix
+ * */
+TransactionManager::TransactionManager(StorageEnv* env) : env_(env) {
+    exec_ = std::make_shared<folly::IOThreadPoolExecutor>(10);
+    interClient_ = std::make_unique<storage::InternalStorageClient>(
+                            exec_,
+                            env_->metaClient_);
+}
+
+/*
+ * multi edges have same local partition and same remote partition
+ * will process as a batch
+ * *Important**Important**Important*
+ * normally, ver will be 0
+ * */
+folly::Future<cpp2::ErrorCode> TransactionManager::addSamePartEdges(
+    size_t vIdLen,
+    GraphSpaceID spaceId,
+    PartitionID localPart,
+    PartitionID remotePart,
+    std::vector<KV>& localEdges,
+    AddEdgesProcessor* processor,
+    folly::Optional<GetBatchFunc> optBatchGetter) {
+    int64_t txnId = TransactionUtils::getSnowFlakeUUID();
+    if (FLAGS_trace_toss) {
+        for (auto& kv : localEdges) {
+            auto srcId = NebulaKeyUtils::getSrcId(vIdLen, kv.first);
+            auto dstId = NebulaKeyUtils::getDstId(vIdLen, kv.first);
+            LOG(INFO) << "begin txn srcId=" << folly::hexlify(srcId)
+                << ", dstId=" << folly::hexlify(dstId) << ", txnId=" << txnId;
+        }
+    }
+    // steps 1: lock edges in memory
+    bool setMemoryLock = true;
+    std::for_each(localEdges.begin(), localEdges.end(), [&](auto& kv){
+        auto keyWoVer = NebulaKeyUtils::keyWithNoVersion(kv.first).str();
+        if (!memLock_.insert(std::make_pair(keyWoVer, txnId)).second) {
+            setMemoryLock = false;
+        }
+    });
+
+    auto cleanup = [=]{
+        for (auto& kv : localEdges) {
+            auto keyWoVer = NebulaKeyUtils::keyWithNoVersion(kv.first).str();
+            auto cit = memLock_.find(keyWoVer);
+            if (cit != memLock_.end() && cit->second == txnId) {
+                memLock_.erase(keyWoVer);
+            }
+        }
+    };
+
+    if (!setMemoryLock) {
+        LOG(ERROR) << "set memory lock failed, txnId=" << txnId;
+        cleanup();
+        return folly::makeFuture(cpp2::ErrorCode::E_MUTATE_EDGE_CONFLICT);
+    } else {
+        LOG_IF(INFO, FLAGS_trace_toss) << "set memory lock succeeded, txnId=" << txnId;
+    }
+
+    // steps 2: batch commit persist locks
+    std::string batch;
+    std::vector<KV> lockData = localEdges;
+    if (!optBatchGetter) {
+        std::transform(lockData.begin(), lockData.end(), lockData.begin(), [=](auto& kv) {
+            if (processor) {
+                std::vector<KV> data{std::make_pair(kv.first, kv.second)};
+                auto optVal = processor->addEdges(localPart, data);
+                std::string val = optVal ? *optVal : "";
+                return std::make_pair(NebulaKeyUtils::toLockKey(kv.first), val);
```

Comment:
We don't need to pass `processor` if there are no indexes in `AddEdgesAtomicProcessor.cpp`. BTW, better return error code when optVal is empty.

Reply:
Never mind.